packages/beacon-node/src/chain/shufflingCache.ts

```diff
@@ -152,39 +116,94 @@ export class ShufflingCache {
+      this.metrics?.shufflingCache.hit.inc();
       return cacheItem.shuffling;
     } else {
-      // promise
+      this.metrics?.shufflingCache.shufflingPromiseNotResolved.inc();
       return cacheItem.promise;
     }
   }
 
   /**
-   * Same to get() function but synchronous.
+   * Gets a cached shuffling via the epoch and decision root.  If the shuffling is not
+   * available it will build it synchronously and return the shuffling.
+   *
+   * NOTE: If a shuffling is already queued and not calculated it will build and resolve
+   * the promise but the already queued build will happen at some later time
    */
-  getSync(shufflingEpoch: Epoch, decisionRootHex: RootHex): EpochShuffling | null {
-    const cacheItem = this.itemsByDecisionRootByEpoch.getOrDefault(shufflingEpoch).get(decisionRootHex);
-    if (cacheItem === undefined) {
-      return null;
+  getSync<T extends ShufflingBuildProps | undefined>(
+    epoch: Epoch,
+    decisionRoot: RootHex,
+    buildProps?: T
+  ): T extends ShufflingBuildProps ? EpochShuffling : EpochShuffling | null {
+    const cacheItem = this.itemsByDecisionRootByEpoch.getOrDefault(epoch).get(decisionRoot);
+    if (!cacheItem) {
+      this.metrics?.shufflingCache.miss.inc();
+    } else if (isShufflingCacheItem(cacheItem)) {
+      this.metrics?.shufflingCache.hit.inc();
+      return cacheItem.shuffling;
+    } else if (buildProps) {
+      // TODO: (@matthewkeil) This should possible log a warning??
+      this.metrics?.shufflingCache.shufflingPromiseNotResolvedAndThrownAway.inc();
+    } else {
+      this.metrics?.shufflingCache.shufflingPromiseNotResolved.inc();
     }
 
-    if (isShufflingCacheItem(cacheItem)) {
-      return cacheItem.shuffling;
+    let shuffling: EpochShuffling | null = null;
+    if (buildProps) {
+      const timer = this.metrics?.shufflingCache.shufflingCalculationTime.startTimer();
+      shuffling = computeEpochShuffling(buildProps.state, buildProps.activeIndices, epoch);
+      timer?.();
+      this.set(shuffling, decisionRoot);
     }
+    return shuffling as T extends ShufflingBuildProps ? EpochShuffling : EpochShuffling | null;
+  }
 
-    // ignore promise
-    return null;
+  /**
+   * Queue asynchronous build for an EpochShuffling
+   */
+  build(epoch: number, decisionRoot: string, state: BeaconStateAllForks, activeIndices: ValidatorIndex[]): void {
+    this.insertPromise(epoch, decisionRoot);
+    /**
+     * TODO: (@matthewkeil) This will get replaced by a proper build queue and a worker to do calculations
+     * on a NICE thread with a rust implementation
+     */
+    setTimeout(() => {
+      const timer = this.metrics?.shufflingCache.shufflingCalculationTime.startTimer();
+      const shuffling = computeEpochShuffling(state, activeIndices, epoch);
+      timer?.();
+      this.set(shuffling, decisionRoot);
+      // wait until after the first slot to help with attestation and block proposal performance
+    }, this.minTimeDelayToBuild);
```

Comment:
why do we have to delay here? this is called at the last 1/3 slot of epoch which is a perfect time to call this heavy operation, we should do it asap

Reply:
I was thinking we might want to delay calculation of the shuffling until the second slot of the epoch because we have trouble with block proposals and attestations in the first slot.  I was guessing that the likelihood of the validator calling for next epoch duties in the first slot was low so the next shuffling wouldn't be needed immediately.

Currently the default for the delay is 5ms so the timeout will be over before the rest of the epoch transition finishes and the synchronous code in `state-transition` transfers thread availability to the `beacon-node`.  The value is configurable via the CLI so we can play with this though to see if/how it affects things.  If you want i can set the default to 0ms but i think the node-defined min for setTimeout is 5ms though.

Reply:
validator does want to have duties in the next epoch https://github.com/ChainSafe/lodestar/blob/4ea7edd6a7482f3131a03f9cde931bf835ce0e03/packages/validator/src/services/attestationDuties.ts#L189